### PR TITLE
[7.x] Add subquery support for aggregate functions

### PR DESCRIPTION
**Pull request by sebdesign**
This PR adds the ability to use subqueries in aggregate methods like `count()`, `sum()`, `min()`, `max()`, `avg()`, and `average()`.

Here are a few examples:

### How many users are logged in actually

```php
$activeUsers = User::count(function () {
    $expiration = now()->subMinutes(config('session.lifetime'));

    Login::selectRaw(1)
          ->where('created_at', '>', $expiration)
          ->whereColumn('user_id', 'users.id')
          ->limit(1);
});
```

#### Generated SQL:
```sql
select count(
    select 1 from `logins`
    where `logins`.`created_at` = ?
    and `user_id` = `users`.`id`
    limit 1
) as aggregate from `users`
```

### Say we want to know the number of views for the videos for the top user

```php
$mostViewsByUser = User::max(function ($query) {
    $query->selectRaw('sum(views)')
        ->from('videos')
        ->whereColumn('user_id', 'users.id');
});
```

#### Generated SQL:

```sql
select max(
    select sum(`views`) from `videos`
    where `user_id` = `users`.`id`
) as aggregate from `users`
```

### Say we want to know the average number of videos the users have published this year

```php
$yearlyAverageVideosByUser = User::avg(function ($query) {
    $query->selectRaw('count(*)')
        ->from('videos')
        ->whereYear('published_at', 2020)
        ->whereColumn('user_id', 'users.id');
});
```

#### Generated SQL:

```sql
select avg(
    select count(*) from `videos`
    where year(`published_at`) = ?
    and `user_id` = `users`.`id`
) as aggregate from `users`
```

These methods accept `Closure`, `Illuminate\Database\Eloquent\Builder`, `Illuminate\Database\Query\Builder`, `Illuminate\Database\Query\Expression` instances as well as strings.

Passing `Expression` instances was still possible before but it was never tested.

Could @reinink or @staudenmeir check if this is looking good?

### CODE REVIEW

File: src/Illuminate/Database/Query/Builder.php [hunk 1]
```diff
@@ -2485,7 +2486,7 @@ public function doesntExistOr(Closure $callback)
     /**
      * Retrieve the "count" result of the query.
      *
-     * @param  string  $columns
+     * @param  array|Closure|\Illuminate\Database\Query\Builder|string  $columns
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php:2489]: Missing leading slash `\Closure`. We always use fully qualified names.

File: src/Illuminate/Database/Query/Builder.php [hunk 2]
```diff
@@ -2496,7 +2497,7 @@ public function count($columns = '*')
     /**
      * Retrieve the minimum value of a given column.
      *
-     * @param  string  $column
+     * @param  Closure|\Illuminate\Database\Query\Builder|string  $column
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php:2500]: same

File: src/Illuminate/Database/Query/Builder.php [hunk 3]
```diff
@@ -2507,7 +2508,7 @@ public function min($column)
     /**
      * Retrieve the maximum value of a given column.
      *
-     * @param  string  $column
+     * @param  Closure|\Illuminate\Database\Query\Builder|string  $column
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php:2511]: etc

**Review by GrahamCampbell — CHANGES_REQUESTED** [review]

File: src/Illuminate/Database/Query/Builder.php [hunk 4]
```diff
@@ -2562,7 +2563,7 @@ public function aggregate($function, $columns = ['*'])
         $results = $this->cloneWithout($this->unions ? [] : ['columns'])
                         ->cloneWithoutBindings($this->unions ? [] : ['select'])
                         ->setAggregate($function, $columns)
-                        ->get($columns);
```
Comment (GrahamCampbell) [inline src/Illuminate/Database/Query/Builder.php:2565]: Why is this removed? I'm sure this has come up before? // cc @themsaid

Comment (sebdesign) [reply src/Illuminate/Database/Query/Builder.php:2565]: I removed it because the get method cannot handle subqueries as columns.
I'm not sure why it's needed though, this argument exists since the first commit ever, but the implementation of the `aggregate` method has changed a lot (mostly cloning issues).

I don't believe it's needed, since `runPaginationCountQuery` executes a similar query without this argument on the `get` method.

**Comment by staudenmeir** [comment]
Please add the executed SQL to your examples.

**Comment by sebdesign** [comment]
@staudenmeir I've updated the PR description with the SQL.

**Comment by sebdesign** [comment]
After inspecting the generated queries of the `Builder::aggregate()` method, I noticed that there is no `limit 1` clause on the queries, but we are only fetching the first result `$results[0]` after. Is there a reason for not limiting the queries?

**Comment by sebdesign** [comment]
But I'm not sure that is the best or more efficient way to calculate those aggregates. This PR was inspired by https://reinink.ca/articles/calculating-totals-in-laravel-using-conditional-aggregates and https://github.com/dwightwatson/aggregate.

**Comment by staudenmeir** [comment]
These queries are actually invalid, it's not possible to use subqueries like this:
https://www.db-fiddle.com/f/eovQxg963yjczFNCPrHiV7/0

**Comment by sebdesign** [comment]
You're right, I think the subquery should be in the `from` clause, like this: https://www.db-fiddle.com/f/7iTrCJqYdSSKi7jKiqFgHp/1

File: src/Illuminate/Database/Query/Builder.php (hunk 4 above)
Comment (taylorotwell) [reply src/Illuminate/Database/Query/Builder.php:2565]: I feel like I can basically guarantee if it is present in the code base it is needed.

**Comment by sebdesign** [comment]
Since these queries are invalid, I'm closing this and I'll submit another PR if I can find a simple solution.
